In [22]:
import pandas as pd
import numpy as np

loc = "MERGED2016_17_PP.csv"

dfraw = pd.read_csv(loc)

C:\Users\isaac\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:2785: DtypeWarning: Columns (6,9,1725,1726,1727,1728,1729,1743,1815,1816,1817,1818,1823,1824,1830,1831) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [23]:
dfraw.head()

UNITID    OPEID  OPEID6                               INSTNM        CITY  \
0  100654   100200    1002             Alabama A & M University      Normal   
1  100663   105200    1052  University of Alabama at Birmingham  Birmingham   
2  100690  2503400   25034                   Amridge University  Montgomery   
3  100706   105500    1055  University of Alabama in Huntsville  Huntsville   
4  100724   100500    1005             Alabama State University  Montgomery   

  STABBR         ZIP                                       ACCREDAGENCY  \
0     AL       35762  Southern Association of Colleges and Schools C...   
1     AL  35294-0110  Southern Association of Colleges and Schools C...   
2     AL  36117-3553  Southern Association of Colleges and Schools C...   
3     AL       35899  Southern Association of Colleges and Schools C...   
4     AL  36104-0271  Southern Association of Colleges and Schools C...   

                     INSTURL  \
0              www.aamu.edu/   
1                www.uab.edu   
2  www.amridgeuniversity.edu   
3                www.uah.edu   
4              www.alasu.edu   

                                              NPCURL    ...      \
0      www2.aamu.edu/scripts/netpricecalc/npcalc.htm    ...       
1           uab.studentaidcalculator.com/survey.aspx    ...       
2                   www2.amridgeuniversity.edu:9091/    ...       
3                                    finaid.uah.edu/    ...       
4  www.alasu.edu/cost-aid/forms/calculator/index....    ...       

   D150_4_LOANNOPELL  C150_L4_LOANNOPELL  D150_L4_LOANNOPELL  \
0              195.0                 NaN                 NaN   
1              229.0                 NaN                 NaN   
2                NaN                 NaN                 NaN   
3              135.0                 NaN                 NaN   
4               55.0                 NaN                 NaN   

   C150_4_NOLOANNOPELL  D150_4_NOLOANNOPELL  C150_L4_NOLOANNOPELL  \
0               0.3258                 89.0                   NaN   
1               0.5603                630.0                   NaN   
2               0.5000                  2.0                   NaN   
3               0.5606                264.0                   NaN   
4               0.2277                101.0                   NaN   

   D150_L4_NOLOANNOPELL  GT_28K_P10  GT_28K_P8  GT_28K_P6  
0                   NaN         NaN        NaN        NaN  
1                   NaN         NaN        NaN        NaN  
2                   NaN         NaN        NaN        NaN  
3                   NaN         NaN        NaN        NaN  
4                   NaN         NaN        NaN        NaN  

[5 rows x 1847 columns]

In [24]:
dfraw.shape

(7175, 1847)

### There are a ridiculous number of columns here, so lets grab some interesting variables and put them in their own dataframe

In [25]:
df = dfraw[["C100_4", "NPT4_PUB"]]
df.columns = ["completion", "public_school_tuition"]

In [26]:
df.head()

completion  public_school_tuition
0      0.0643                15567.0
1      0.3018                16475.0
2      0.1429                    NaN
3      0.1533                19423.0
4      0.0725                15037.0

In [27]:
df.corr()

completion  public_school_tuition
completion               1.000000               0.515731
public_school_tuition    0.515731               1.000000

### Joy suggested that INEXPFTE (instructional expenditures per full-time student) and ADM_RATE (admissions rate) are correlated with completion rates, so lets add those to the dataframe and run the correlation

In [28]:
df["expenditures_per_student"] = dfraw["INEXPFTE"]
df["adm_rate"] = dfraw["ADM_RATE"]

C:\Users\isaac\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
C:\Users\isaac\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


### I suggest looking at pell_grant_rate, and loan_principal. The intuition is that if you qualify for pell you're from a disadvantaged background and less likely to graduate

In [29]:
df["pell_grant_rate"]=dfraw["PCTPELL"]

C:\Users\isaac\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


In [30]:
df.head()

completion  public_school_tuition  expenditures_per_student  adm_rate  \
0      0.0643                15567.0                    5888.0    0.8738   
1      0.3018                16475.0                   17055.0    0.5814   
2      0.1429                    NaN                    5417.0       NaN   
3      0.1533                19423.0                   10558.0    0.7628   
4      0.0725                15037.0                    7889.0    0.4590   

   pell_grant_rate  
0           0.7039  
1           0.3525  
2           0.6971  
3           0.2949  
4           0.7815

### Another place to look for impact on completion rates is the share of degree seeking students who are of a particular race or ethnicity

In [31]:
df["white"] = dfraw["UGDS_WHITE"]
df["black"] = dfraw["UGDS_BLACK"]
df["hispanic"] = dfraw["UGDS_HISP"]
df["asian"] = dfraw["UGDS_ASIAN"]
df["american_indian"] = dfraw["UGDS_AIAN"]
df["hawaiian_pcislander"] = dfraw["UGDS_NHPI"]
df["race_ethnicity_two"] =dfraw["UGDS_2MOR"]
df["nonresident"] = dfraw["UGDS_NRA"]
df["unknown"] = dfraw["UGDS_UNKN"]

C:\Users\isaac\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
C:\Users\isaac\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
C:\Users\isaac\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.

In [32]:
df.corr()

completion  public_school_tuition  \
completion                  1.000000               0.515731   
public_school_tuition       0.515731               1.000000   
expenditures_per_student    0.459406               0.292337   
adm_rate                   -0.306999               0.062488   
pell_grant_rate            -0.569687              -0.072446   
white                       0.278313               0.202425   
black                      -0.291535               0.003498   
hispanic                   -0.149864              -0.235117   
asian                       0.263919               0.042371   
american_indian            -0.042858              -0.107501   
hawaiian_pcislander        -0.072222              -0.077829   
race_ethnicity_two          0.171952               0.018126   
nonresident                 0.176216               0.204074   
unknown                    -0.166568              -0.013419   

                          expenditures_per_student  adm_rate  pell_grant_rate  \
completion                                0.459406 -0.306999        -0.569687   
public_school_tuition                     0.292337  0.062488        -0.072446   
expenditures_per_student                  1.000000 -0.253013        -0.282385   
adm_rate                                 -0.253013  1.000000         0.145925   
pell_grant_rate                          -0.282385  0.145925         1.000000   
white                                     0.105862  0.143278        -0.338549   
black                                    -0.120008 -0.048812         0.355269   
hispanic                                 -0.128894 -0.008082         0.230680   
asian                                     0.116228 -0.201542        -0.163218   
american_indian                           0.019564  0.034859         0.008184   
hawaiian_pcislander                      -0.018091  0.033931         0.061994   
race_ethnicity_two                        0.087839 -0.069470        -0.144701   
nonresident                               0.221100 -0.182576        -0.278088   
unknown                                  -0.012070 -0.022053        -0.025942   

                             white     black  hispanic     asian  \
completion                0.278313 -0.291535 -0.149864  0.263919   
public_school_tuition     0.202425  0.003498 -0.235117  0.042371   
expenditures_per_student  0.105862 -0.120008 -0.128894  0.116228   
adm_rate                  0.143278 -0.048812 -0.008082 -0.201542   
pell_grant_rate          -0.338549  0.355269  0.230680 -0.163218   
white                     1.000000 -0.487208 -0.565485 -0.225557   
black                    -0.487208  1.000000 -0.208349 -0.126757   
hispanic                 -0.565485 -0.208349  1.000000  0.013849   
asian                    -0.225557 -0.126757  0.013849  1.000000   
american_indian          -0.096462 -0.092085 -0.062749 -0.053453   
hawaiian_pcislander      -0.104106 -0.039819 -0.017736  0.103119   
race_ethnicity_two        0.007263 -0.106087 -0.099518  0.095215   
nonresident              -0.091099 -0.121283 -0.064033  0.159135   
unknown                  -0.173129 -0.032272 -0.091325 -0.016698   

                          american_indian  hawaiian_pcislander  \
completion                      -0.042858            -0.072222   
public_school_tuition           -0.107501            -0.077829   
expenditures_per_student         0.019564            -0.018091   
adm_rate                         0.034859             0.033931   
pell_grant_rate                  0.008184             0.061994   
white                           -0.096462            -0.104106   
black                           -0.092085            -0.039819   
hispanic                        -0.062749            -0.017736   
asian                           -0.053453             0.103119   
american_indian                  1.000000            -0.004183   
hawaiian_pcislander             -0.004183             1.000000   
race_ethnicity_two               0.007327          

### These results are promising so far, let's remove NaN observations and run a regression

In [33]:
df.dropna(inplace = True)
df.head()

C:\Users\isaac\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.


completion  public_school_tuition  expenditures_per_student  adm_rate  \
0      0.0643                15567.0                    5888.0    0.8738   
1      0.3018                16475.0                   17055.0    0.5814   
3      0.1533                19423.0                   10558.0    0.7628   
4      0.0725                15037.0                    7889.0    0.4590   
5      0.4385                21676.0                   10478.0    0.5259   

   pell_grant_rate   white   black  hispanic   asian  american_indian  \
0           0.7039  0.0256  0.9129    0.0076  0.0019           0.0024   
1           0.3525  0.5786  0.2626    0.0309  0.0598           0.0028   
3           0.2949  0.7148  0.1131    0.0411  0.0414           0.0120   
4           0.7815  0.0138  0.9337    0.0111  0.0028           0.0013   
5           0.1938  0.7841  0.1037    0.0437  0.0118           0.0036   

   hawaiian_pcislander  race_ethnicity_two  nonresident  unknown  
0               0.0017              0.0401       0.0065   0.0013  
1               0.0004              0.0387       0.0179   0.0083  
3               0.0000              0.0181       0.0303   0.0292  
4               0.0004              0.0111       0.0159   0.0100  
5               0.0009              0.0297       0.0192   0.0033

### Next let's run a regression and see what the results are

In [36]:
import statsmodels.formula.api as sm

reg1 = sm.ols(formula='completion ~ public_school_tuition + expenditures_per_student + adm_rate + pell_grant_rate + white + black + hispanic + asian + american_indian + hawaiian_pcislander + race_ethnicity_two + nonresident + unknown', data=df).fit()
reg1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             completion   R-squared:                       0.550
Model:                            OLS   Adj. R-squared:                  0.539
Method:                 Least Squares   F-statistic:                     49.81
Date:                Fri, 19 Oct 2018   Prob (F-statistic):           4.54e-83
Time:                        23:36:59   Log-Likelihood:                 401.28
No. Observations:                 543   AIC:                            -774.6
Df Residuals:                     529   BIC:                            -714.4
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
============================================================================================
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
Intercept                    3.1136     56.780      0.055      0.956    -108.427     114.655
public_school_tuition     1.328e-05    1.5e-06      8.831      0.000    1.03e-05    1.62e-05
expenditures_per_student  1.121e-05   1.42e-06      7.901      0.000    8.42e-06     1.4e-05
adm_rate                    -0.1404      0.034     -4.124      0.000      -0.207      -0.074
pell_grant_rate             -0.3142      0.068     -4.599      0.000      -0.448      -0.180
white                       -2.8889     56.781     -0.051      0.959    -114.433     108.655
black                       -3.0002     56.783     -0.053      0.958    -114.548     108.548
hispanic                    -2.8637     56.782     -0.050      0.960    -114.409     108.682
asian                       -2.6588     56.787     -0.047      0.963    -114.214     108.897
american_indian             -3.2026     56.777     -0.056      0.955    -114.739     108.334
hawaiian_pcislander         -5.9131     56.823     -0.104      0.917    -117.540     105.714
race_ethnicity_two          -3.1044     56.779     -0.055      0.956    -114.645     108.436
nonresident                 -2.9100     56.776     -0.051      0.959    -114.444     108.624
unknown                     -2.9460     56.784     -0.052      0.959    -114.495     108.603
==============================================================================
Omnibus:                        7.250   Durbin-Watson:                   1.616
Prob(Omnibus):                  0.027   Jarque-Bera (JB):                9.793
Skew:                           0.103   Prob(JB):                      0.00747
Kurtosis:                       3.625   Cond. No.                     6.35e+08
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 6.35e+08. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### Our race/ethnicity variables leave something to be desired, lets remove some of them and see if the significance increases

In [37]:
reg2 = sm.ols(formula='completion ~ public_school_tuition + expenditures_per_student + adm_rate + pell_grant_rate + white + black + hispanic + asian', data=df).fit()
reg2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             completion   R-squared:                       0.540
Model:                            OLS   Adj. R-squared:                  0.533
Method:                 Least Squares   F-statistic:                     78.44
Date:                Fri, 19 Oct 2018   Prob (F-statistic):           3.99e-85
Time:                        23:38:51   Log-Likelihood:                 395.26
No. Observations:                 543   AIC:                            -772.5
Df Residuals:                     534   BIC:                            -733.8
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
============================================================================================
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
Intercept                   -0.0138      0.082     -0.168      0.866      -0.175       0.147
public_school_tuition     1.344e-05    1.5e-06      8.934      0.000    1.05e-05    1.64e-05
expenditures_per_student  1.167e-05   1.41e-06      8.260      0.000    8.89e-06    1.44e-05
adm_rate                    -0.1529      0.034     -4.493      0.000      -0.220      -0.086
pell_grant_rate             -0.3355      0.068     -4.950      0.000      -0.469      -0.202
white                        0.2569      0.079      3.265      0.001       0.102       0.411
black                        0.1534      0.084      1.823      0.069      -0.012       0.319
hispanic                     0.2881      0.082      3.532      0.000       0.128       0.448
asian                        0.4472      0.131      3.409      0.001       0.189       0.705
==============================================================================
Omnibus:                        6.186   Durbin-Watson:                   1.607
Prob(Omnibus):                  0.045   Jarque-Bera (JB):                8.013
Skew:                           0.090   Prob(JB):                       0.0182
Kurtosis:                       3.567   Cond. No.                     6.61e+05
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 6.61e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### Narrowing down the specification in the race/ethnicity category seems to have improved the predictive capabilities of our model. Race/ethnicity factors are now signficant except for the black variable

### From the beginning we had a problem with our money-based variables. The coefficients are very small. So lets consider them in an either thousands of dollars specifcation or a log specification

In [38]:
df["log_public_school_tuition"] = np.log(df['public_school_tuition'])
df["public_school_tuition_thousands"] = df["public_school_tuition"]/1000
df["log_expenditures_per_student"] = np.log(df['expenditures_per_student'])
df["expenditures_per_student_thousands"] = df["expenditures_per_student"]/1000

C:\Users\isaac\Anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
C:\Users\isaac\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
C:\Users\isaac\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.

In [39]:
df.head()

completion  public_school_tuition  expenditures_per_student  adm_rate  \
0      0.0643                15567.0                    5888.0    0.8738   
1      0.3018                16475.0                   17055.0    0.5814   
3      0.1533                19423.0                   10558.0    0.7628   
4      0.0725                15037.0                    7889.0    0.4590   
5      0.4385                21676.0                   10478.0    0.5259   

   pell_grant_rate   white   black  hispanic   asian  american_indian  \
0           0.7039  0.0256  0.9129    0.0076  0.0019           0.0024   
1           0.3525  0.5786  0.2626    0.0309  0.0598           0.0028   
3           0.2949  0.7148  0.1131    0.0411  0.0414           0.0120   
4           0.7815  0.0138  0.9337    0.0111  0.0028           0.0013   
5           0.1938  0.7841  0.1037    0.0437  0.0118           0.0036   

   hawaiian_pcislander  race_ethnicity_two  nonresident  unknown  \
0               0.0017              0.0401       0.0065   0.0013   
1               0.0004              0.0387       0.0179   0.0083   
3               0.0000              0.0181       0.0303   0.0292   
4               0.0004              0.0111       0.0159   0.0100   
5               0.0009              0.0297       0.0192   0.0033   

   log_public_school_tuition  public_school_tuition_thousands  \
0                   9.652909                           15.567   
1                   9.709599                           16.475   
3                   9.874213                           19.423   
4                   9.618269                           15.037   
5                   9.983961                           21.676   

   log_expenditures_per_student  expenditures_per_student_thousands  
0                      8.680672                               5.888  
1                      9.744199                              17.055  
3                      9.264639                              10.558  
4                      8.973225                               7.889  
5                      9.257033                              10.478

### Let's run the model with log specifications. The interpretation is that for each percent increase in costs/tuition, the completion rate should be effected by some percentage of 100.

In [40]:
reg3 = sm.ols(formula='completion ~ log_public_school_tuition + log_expenditures_per_student + adm_rate + pell_grant_rate + white + black + hispanic + asian', data=df).fit()
reg3.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             completion   R-squared:                       0.505
Model:                            OLS   Adj. R-squared:                  0.498
Method:                 Least Squares   F-statistic:                     68.11
Date:                Fri, 19 Oct 2018   Prob (F-statistic):           1.19e-76
Time:                        23:41:05   Log-Likelihood:                 375.21
No. Observations:                 543   AIC:                            -732.4
Df Residuals:                     534   BIC:                            -693.8
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
================================================================================================
                                   coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------
Intercept                       -2.0631      0.234     -8.817      0.000      -2.523      -1.603
log_public_school_tuition        0.1191      0.019      6.123      0.000       0.081       0.157
log_expenditures_per_student     0.1309      0.016      8.108      0.000       0.099       0.163
adm_rate                        -0.1532      0.035     -4.364      0.000      -0.222      -0.084
pell_grant_rate                 -0.3880      0.070     -5.538      0.000      -0.526      -0.250
white                            0.3061      0.081      3.764      0.000       0.146       0.466
black                            0.2096      0.087      2.416      0.016       0.039       0.380
hispanic                         0.3615      0.085      4.278      0.000       0.195       0.528
asian                            0.5256      0.135      3.880      0.000       0.260       0.792
==============================================================================
Omnibus:                       18.232   Durbin-Watson:                   1.580
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               29.367
Skew:                           0.250   Prob(JB):                     4.20e-07
Kurtosis:                       4.024   Cond. No.                         613.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### The log spec dropped the r^2, but had easily interperable coefficients

### Let's try the thousands spec

In [41]:
import statsmodels.formula.api as sm

reg4 = sm.ols(formula='completion ~ public_school_tuition_thousands + expenditures_per_student_thousands + adm_rate + pell_grant_rate + white + black + hispanic + asian', data=df).fit()
reg4.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             completion   R-squared:                       0.540
Model:                            OLS   Adj. R-squared:                  0.533
Method:                 Least Squares   F-statistic:                     78.44
Date:                Fri, 19 Oct 2018   Prob (F-statistic):           3.99e-85
Time:                        23:41:48   Log-Likelihood:                 395.26
No. Observations:                 543   AIC:                            -772.5
Df Residuals:                     534   BIC:                            -733.8
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
======================================================================================================
                                         coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------
Intercept                             -0.0138      0.082     -0.168      0.866      -0.175       0.147
public_school_tuition_thousands        0.0134      0.002      8.934      0.000       0.010       0.016
expenditures_per_student_thousands     0.0117      0.001      8.260      0.000       0.009       0.014
adm_rate                              -0.1529      0.034     -4.493      0.000      -0.220      -0.086
pell_grant_rate                       -0.3355      0.068     -4.950      0.000      -0.469      -0.202
white                                  0.2569      0.079      3.265      0.001       0.102       0.411
black                                  0.1534      0.084      1.823      0.069      -0.012       0.319
hispanic                               0.2881      0.082      3.532      0.000       0.128       0.448
asian                                  0.4472      0.131      3.409      0.001       0.189       0.705
==============================================================================
Omnibus:                        6.186   Durbin-Watson:                   1.607
Prob(Omnibus):                  0.045   Jarque-Bera (JB):                8.013
Skew:                           0.090   Prob(JB):                       0.0182
Kurtosis:                       3.567   Cond. No.                         663.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Here the r^2 remained higher. So it probably makes more sense to keep the thousands specification. Here the coefficients are much more interperable than before. For every thousand dollars more the school charges in tuition, the likelihood of degree completion increases by 1.21%. The same can be said of the impact of an extra thousand dollars in expenditures.

### Next lets see whether the r^2 improves without an intercept

In [43]:
reg5 = sm.ols(formula='completion ~ public_school_tuition_thousands + expenditures_per_student_thousands + adm_rate + pell_grant_rate + white + black + hispanic + asian - 1', data=df).fit()
reg5.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             completion   R-squared:                       0.877
Model:                            OLS   Adj. R-squared:                  0.875
Method:                 Least Squares   F-statistic:                     476.7
Date:                Fri, 19 Oct 2018   Prob (F-statistic):          8.18e-238
Time:                        23:44:44   Log-Likelihood:                 395.24
No. Observations:                 543   AIC:                            -774.5
Df Residuals:                     535   BIC:                            -740.1
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
======================================================================================================
                                         coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------------------
public_school_tuition_thousands        0.0134      0.001      9.024      0.000       0.010       0.016
expenditures_per_student_thousands     0.0116      0.001      8.647      0.000       0.009       0.014
adm_rate                              -0.1548      0.032     -4.838      0.000      -0.218      -0.092
pell_grant_rate                       -0.3386      0.065     -5.198      0.000      -0.467      -0.211
white                                  0.2456      0.042      5.895      0.000       0.164       0.327
black                                  0.1434      0.059      2.413      0.016       0.027       0.260
hispanic                               0.2778      0.054      5.172      0.000       0.172       0.383
asian                                  0.4324      0.097      4.449      0.000       0.241       0.623
==============================================================================
Omnibus:                        6.131   Durbin-Watson:                   1.605
Prob(Omnibus):                  0.047   Jarque-Bera (JB):                7.940
Skew:                           0.089   Prob(JB):                       0.0189
Kurtosis:                       3.565   Cond. No.                         411.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### So far our model has eight independent variables. The r^2 is high and this is a predictive model we can leave it for now. Our next step can be testing the model specification with a split test. To do this we first need to employ a regression method that is amenable to a split test. Let's load some libraries

In [44]:
%matplotlib inline

import scipy.stats as stats
import matplotlib.pyplot as plt
import sklearn

In [49]:
from sklearn.linear_model import LinearRegression
X = df.drop(['completion', "public_school_tuition", "expenditures_per_student", "log_public_school_tuition", "log_expenditures_per_student", "american_indian", "hawaiian_pcislander" ,"race_ethnicity_two", "nonresident", "unknown"], axis = 1)

lm = LinearRegression(fit_intercept=False)
lm

LinearRegression(copy_X=True, fit_intercept=False, n_jobs=1, normalize=False)

In [50]:
lm.fit(X, df.completion)

LinearRegression(copy_X=True, fit_intercept=False, n_jobs=1, normalize=False)

### Let's check some coefficients to be sure they match with our previous model

In [54]:
reg5.params

public_school_tuition_thousands       0.013397
expenditures_per_student_thousands    0.011595
adm_rate                             -0.154828
pell_grant_rate                      -0.338602
white                                 0.245630
black                                 0.143413
hispanic                              0.277766
asian                                 0.432387
dtype: float64

In [55]:
pd.DataFrame(list(zip(X.columns, lm.coef_)), columns = ["features", 'estimated coefficients'])

features  estimated coefficients
0                            adm_rate               -0.154828
1                     pell_grant_rate               -0.338602
2                               white                0.245630
3                               black                0.143413
4                            hispanic                0.277766
5                               asian                0.432387
6     public_school_tuition_thousands                0.013397
7  expenditures_per_student_thousands                0.011595

### Great our models match, that's a good start, let's split test

In [56]:
X_train, X_test, Y_train, Y_test = sklearn.model_selection.train_test_split(
    X, df.completion, test_size=.33, random_state = 5)
print(X_train.shape)
print(X_test.shape)
print(Y_train.shape)
print(Y_test.shape)

(363, 8)
(180, 8)
(363,)
(180,)


In [57]:
lm=LinearRegression()
lm.fit(X_train, Y_train)
pred_train = lm.predict(X_train)
pred_test = lm.predict(X_test)

In [58]:
print('Fit a model X_train, and calculate MSE with Y_train:', np.mean((Y_train - lm.predict(X_train)) ** 2))
print('Fit a model X_train, and calculate MSE with X_test, Y_test:', np.mean((Y_test - lm.predict(X_test)) ** 2))

Fit a model X_train, and calculate MSE with Y_train: 0.013330520311307116
Fit a model X_train, and calculate MSE with X_test, Y_test: 0.014466214287957996


### The mean of the squared errors is very small which I take to be a positive.